File: Problem2.js
// Time Complexity : O(V+E)
// Space Complexity : O(V+E)
// Did this code successfully run on Leetcode : Yes

var canFinish = function (numCourses, prerequisites) {
    let map = new Map();
    let indegrees = new Array(numCourses).fill(0);
    for (let edge of prerequisites) {
        indegrees[edge[0]]++;
        if (!map.has(edge[1])) {
            map.set(edge[1], []);
        }
        map.get(edge[1]).push(edge[0]);
    }
    let q = [];
    let count = 0;
    for (let i = 0; i < indegrees.length; i++) {
        if (indegrees[i] === 0) {
            q.push(i);
            count++;
        }
    }

    if (q.length === 0) return false;

    while (q.length) {
        let current = q.shift();
        let children = map.get(current);
        if (children) {
            for (let child of children) {
                indegrees[child]--;
                if (indegrees[child] === 0) {
                    q.push(child);
                    count++;
                }

            }
        }
    }
    return count === numCourses;
};

File: Problem2.cs
public class CourseSchedule
    {
        // Time Complexity : O(V + E) - V is all vertices/courses and E is all edges/dependencies
        // Space Complexity : O(V + E) - V is all vertices/courses and E is all edges/dependencies and hashmap is used as auxiliary space 
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No
        public bool CanFinish(int numCourses, int[][] prerequisites)
        {
            Dictionary<int, List<int>> map = new Dictionary<int, List<int>>();
            int[] indegrees = new int[numCourses];
            foreach (var pr in prerequisites)
            {
                //        incoming <- outgoing
                //(1,0) :   1      <-   0
                int incoming = pr[0]; //dependent
                int outgoing = pr[1]; //independent

                indegrees[incoming]++;
                if(!map.ContainsKey(outgoing))
                {
                    map.Add(outgoing, new List<int>());
                }
                map[outgoing].Add(incoming);
            }

            int count = 0;
            Queue<int> q = new Queue<int>();

            //first go over indegrees array to get all independent nodes (0) and add it to Queue
            for(int i = 0; i < indegrees.Length; i++)
            {
                if(indegrees[i] == 0)
                {
                    q.Enqueue(i);
                    count++;
                }
            }

            if (count == numCourses) return true; // if everything is independent and no edges

            if (q.Count == 0) return false; // if everything is dependent upon each other nothing goes in queue

            while(q.Count > 0)
            {
                //size not required as we are not calculating how many semesters
                int curr = q.Dequeue();
                if (map.ContainsKey(curr))
                {
                    List<int> dependencies = map[curr];
                    if (dependencies.Count > 0)
                    {
                        foreach (var dependent in dependencies)
                        {
                            indegrees[dependent]--;
                            if (indegrees[dependent] == 0)
                            {
                                q.Enqueue(dependent);
                                count++;
                                if (count == numCourses) return true; //we can finish the BFS early and leave the other elements in queue
                            }
                        }
                    }
                }
            }
            return count == numCourses;
        }
    }
